Sources/NIOHPACK/HPACKDecoder.swift

```diff
@@ -0,0 +1,187 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// An `HPACKDecoder` maintains its own dynamic header table and uses that to
+/// decode indexed HTTP headers, along with Huffman-encoded strings and
+/// HPACK-encoded integers.
+///
+/// Its only shared state is its dynamic header table; each decode operation
+/// will update the header table as described in RFC 7541, appending and
+/// evicting items as described there.
+public struct HPACKDecoder {
+    public static let maxDynamicTableSize = DynamicHeaderTable.defaultSize
+    
+    // private but tests
+    var headerTable: IndexedHeaderTable
+    
+    private let allocator: ByteBufferAllocator
+    
+    private let huffmanDecoder = HuffmanDecoder()
+    
+    var dynamicTableLength: Int {
+        return headerTable.dynamicTableLength
+    }
+    
+    /// The maximum size of the dynamic table. This is defined in RFC 7541
+    /// to be the sum of [name-octet-count] + [value-octet-count] + 32 for
+    /// each header it contains.
+    public internal(set) var maxDynamicTableLength: Int {
+        get { return headerTable.maxDynamicTableLength }
+        /* private but tests */
+        set { headerTable.maxDynamicTableLength = newValue }
+    }
+    
+    /// Creates a new decoder
+    ///
+    /// - Parameter maxDynamicTableSize: Maximum allowed size of the dynamic header table.
+    public init(maxDynamicTableSize: Int = HPACKDecoder.maxDynamicTableSize, allocator: ByteBufferAllocator = ByteBufferAllocator()) {
+        self.headerTable = IndexedHeaderTable(maxDynamicTableSize: maxDynamicTableSize)
+        self.allocator = allocator
+    }
+    
+    /// Reads HPACK encoded header data from a `ByteBuffer`.
+    ///
+    /// It is expected that the buffer cover only the bytes of the header list, i.e.
+    /// that this is in fact a slice containing only the payload bytes of a
+    /// `HEADERS` or `CONTINUATION` frame.
+    ///
+    /// - Parameter buffer: A byte buffer containing the encoded header bytes.
+    /// - Returns: An array of (name, value) pairs.
+    /// - Throws: HpackDecoder.Error in the event of a decode failure.
+    public mutating func decodeHeaders(from buffer: inout ByteBuffer) throws -> HPACKHeaders {
+        // take a local copy to mutate
+        var bufCopy = buffer
+        
+        var decodedBuffer = allocator.buffer(capacity: buffer.capacity)
```

Comment:
I can do that; I'll add a caveat about thread-safety to the type documentation. I'm inclined to go with the second option, but if anything start at PAGE_SIZE or the max dynamic table size, which corresponds to SETTINGS_MAX_HEADER_LIST_SIZE. It can then grow as needed by changes to that value, but not shrink.